lib/persisted-model.js

```diff
@@ -1081,9 +1100,88 @@ PersistedModel.bulkUpdate = function(updates, callback) {
     }
   });
 
-  async.parallel(tasks, callback);
+  async.parallel(tasks, function(err) {
+    if (err) return callback(err);
+    if (conflicts.length) {
+      err = new Error('Conflict');
+      err.statusCode = 409;
+      err.details = { conflicts: conflicts };
+      return callback(err);
+    }
+    callback();
+  });
 };
 
+function applyUpdate(Model, id, data, change, conflicts, cb) {
+  var Change = Model.getChangeModel();
+  Model.findById(id, function(err, inst) {
+    if (err) return cb(err);
+
+    var rev = inst ?
+      Change.revisionForInst(inst) :
+      null;
+
+    if (rev !== change.prev) {
+      debug('Detected non-rectified change of %s %j',
+        Model.modelName, id);
+      debug('\tExpected revision: %s', change.rev);
+      debug('\tActual revision:   %s', rev);
+      conflicts.push(change);
+      return Change.rectifyModelChanges(Model.modelName, [id], cb);
+    }
+
+    // TODO(bajtos) modify `data` so that it instructs
+    // the connector to remove any properties included in "inst"
+    // but not included in `data`
+    // See https://github.com/strongloop/loopback/issues/1215
+
+    Model.updateAll(inst.toObject(), data, function(err, count) {
+      if (err) return cb(err);
+      if (count === 0) {
+        debug('UpdateAll detected non-rectified change of %s %j',
+          Model.modelName, id);
+        conflicts.push(change);
+        // NOTE(bajtos) updateAll triggers change rectification
+        // for all model instances, even when no records were updated,
+        // thus we don't need to rectify explicitly ourselves
+      } else if (count !== 1) {
+        debug('%s.updateAll modified unexpected number of instances: %j',
+          Model.modelName, count);
+        return cb(new Error(
+          'Cannot apply bulk updates, ' +
+          'the connector does not correctly report ' +
+          'the number of updated records.'));
+      }
+      cb();
+    });
+  });
+}
+
+function applyCreate(Model, id, data, change, conflicts, cb) {
+  Model.create(data, function(createErr) {
+    if (!createErr) return cb();
+
+    // We don't have a reliable way how to detect the situation
+    // where he model was not create because of a duplicate id
+    // The workaround is to query the DB to check if the model already exists
+    Model.findById(id, function(findErr, inst) {
```

Comment:
> The juggler should probably take this on though (an optmization or reporting a consistent error).

I have already looked into this, databases error messages/code don't distinguish between a duplicate primary key (e.g. id) and a duplicate value in a unique column, at least not in MySQL and MongoDB.

See e.g. https://github.com/felixge/node-mysql/issues/372:

> Detecting which fields are duplicated is not part of the mysql protocol (afaik), and we're not going to start human readable error strings from the server.